src/Perl6/Metamodel/RolePunning.nqp

```diff
@@ -57,6 +57,24 @@ role Perl6::Metamodel::RolePunning {
         self.pun($obj)
     }
 
+    my $proxy_type := nqp::null();
+    method !make_proxy($cont) {
+        if nqp::isnull($proxy_type) {
+            $proxy_type := nqp::gethllsym('Raku', 'Proxy');
+        }
+
+        my $proxy   := nqp::create($proxy_type);
+        my $updated := 0;
+        nqp::bindattr($proxy, $proxy_type, '&!FETCH', -> $var {
+            $updated ?? nqp::decont($cont) !! $!pun;
+        });
+        nqp::bindattr($proxy, $proxy_type, '&!STORE', -> $var, $val {
+            nqp::assign($cont, $val);
+            $updated := 1;
```

Comment:
I feel very much uncertain about the use of `$updated`. What if `$val` is the same role? You get the same container with the same role inside but "updated".

Reply:
Why would a method store the original role in the invocant container? It would have to extract it from the pun first, so this will hardly be done by accident. And if a method does this for some reason, you just end up with a container that fetches the role, rather than its pun. If you then call another method on the container, that other method's invocant parameter will be bound to a new `Proxy`.

Maybe it helps if you think of `$updated` as `$initialized`. It's not called that way because the latter name would imply that you can only store a concrete thing in the container and the point of the `Proxy` is to allow code to store _anything_ that conforms to the type constraint of the original container. Once something is stored in the container, that's what you will get if you ask for the container's contents. Until something is stored in the container, you'll get its default value which happens to be the pun.

Reply:
> Why would a method store the original role in the invocant container?

Say, something like this:

```raku
method foo(::?ROLE \SELF:) {
    SELF = self.are-we-ready ?? SELF.new !! ::?ROLE;
}
```

Without arguing why exactly this approach, the snippet is not prohibited in general. Though I didn't consider this case thoroughly and may miss something, but the result of `find_method` would be cached and the very same thunk would be invoked for the role. But wether I'm right here is not really relevant at the moment due to the bigger issue.